torch/csrc/autograd/engine.cpp

```diff
@@ -701,22 +695,31 @@ void validate_outputs(
     if (c10::typeMetaToScalarType(metadata.options().dtype()) != grad.scalar_type()) {
       grad = grad.to(c10::typeMetaToScalarType(metadata.options().dtype()));
     }
-    if (grad.device() != metadata.device() &&
-        grad.dim() == 0) {
-      grad = grad.to(metadata.device());
+    if (grad.dtype() != metadata.dtype()) {
+       std::stringstream ss;
+       ss << "invalid gradient at index " << i << " - expected dtype ";
+       ss << metadata.dtype() << " but got " << grad.dtype();
+       AT_ERROR(format_error(ss.str()));
     }
-    if (!is_compatible_type(metadata.options(), grad.options())) {
+    if (grad.layout() != metadata.layout()) {
        std::stringstream ss;
-       ss << "invalid gradient at index " << i << " - expected type ";
-       ss << metadata.options() << " but got " << grad.options();
+       ss << "invalid gradient at index " << i << " - expected layout ";
```

Comment:
There are currently four layouts: sparse, sparsecsr, mkldnn and strided. Currently, the only cross combination of layouts for tensor and grad that's supported is `(Strided, Sparse)` (based on the check that was being done by `is_compatible_type`). I am not sure about mkldn but it seems like we should support all possible combinations of (tensor.layout(), tensor.grad.layout()) for the other three layouts. cc. @albanD

Reply:
I think we can keep this current layout logic as-is and only update the subclass case.